modules/electrophysiology_browser/php/sessions.class.inc

```diff
@@ -448,6 +481,591 @@ class Sessions extends \NDB_Page
+                     * from the parameter table
+                     */
+                    if (empty($values['instance_id'])) {
+                        $parameterID = $db->pselectone(
+                            "SELECT AnnotationParameterID
+                            FROM physiological_annotation_parameter
+                            WHERE AnnotationFileID=:annotationFID",
+                            ['annotationFID' => $json_ID]
+                        );
+                        $instance['AnnotationParameterID'] = $parameterID;
+
+                        $db->insert('physiological_annotation_instance', $instance);
+                    } else {
+                        //Update instance table
+                        if (!empty($values['instance_id'])) {
+                            $db->update(
+                                'physiological_annotation_instance',
+                                $instance,
+                                ['AnnotationInstanceID' => $values['instance_id']]
+                            );
+                        }
+                        //Update parameter table
+                        if (!empty($values['parameter_id'])) {
+                            $db->update(
+                                'physiological_annotation_parameter',
+                                $metadata,
+                                ['AnnotationParameterID' => $values['parameter_id']]
+                            );
+                        }
+
+                    }
+
+                    //In all cases where files are not new,
+                    //set LastUpdate time for all related files
+                    $db->update(
+                        'physiological_annotation_file',
+                        ['LastUpdate' => date("Y-m-d H:i:s")],
+                        ['PhysiologicalFileID' => $physioFileID]
+                    );
+
+                }
+            }
+        }
+    }
+
+    /**
+     * Deletes one annotation from a given physiological file
+     *
+     * @param ServerRequestInterface $request DELETE request from server
+     *
+     * @return void
+     */
+    function _deleteAnnotation(ServerRequestInterface $request)
+    {
+        $values = $request->getQueryParams();
+        $user   = \NDB_Factory::singleton()->user();
+        $db     = \NDB_Factory::singleton()->database();
+        $params = [
+            'AnnotationFileID'     => $values['file_id'],
+            'AnnotationInstanceID' => $values['instance_id']
+        ];
+
+        if ($this->_hasEditDerivPerms($user)) {
+            $db->delete("physiological_annotation_instance", $params);
+        }
+    }
+
+    /**
+     * Updates the derivative files associated with the given
+     * physiological file ID
+     *
+     * @param int $physioFileID File ID of annotation files to update
+     *
+     * @return void
+     * @throws SodiumException
+     */
+    static function updateDerivativeFiles(int $physioFileID): void
```

Comment:
I'm not too involved in the EEG design usually so this comment might be inaccurate.

This function does a ton of SQL selects, some even in a loop of instances, which is a good way to waste a lot of time. Is there no way to compile everything that needs to be queried and then run a fraction of the queries to get all the needed data ?